### DISCUSSION TITLE
Packaging in a Node project gives a cannot find module error

### DISCUSSION THREAD

**Discussion by mikehalleen** [artilleryio/artillery#2347] [Artillery]
Hi,

I have a basic test, and the yml version runs fine when executed as 

`PS C:\Users\USER\artv1> artillery run loadtest.yml`

But now I want to package it as a Node project and it gives this error

`Error: Cannot find module 'C:\Users\USER\artv1\node_modules\artillery\lib'. Please verify that the package.json has a valid "main" entry`

The reason I'm trying a node package is to eventually upload to AWS to run as lamba, which I understand is experimental and may have alternatives, but these errors are happening entirely locally before AWS even enters the picture.  (I do get the same error running on AWS, just with a AWS path rather than my C:\Users... one)

Here are the steps I took:

nvm install -g artillery
[wrote and tested yml script, it works]

mkdir artv1
cd artv1
[copy yml script here]
[create package.json and index.js, see below]
nvm install artillery 
node index.js 

```
node:internal/modules/cjs/loader:452
      throw err;
      ^


Error: Cannot find module 'C:\Users\USER\artv1\node_modules\artillery\lib'. Please verify that the package.json has a valid "main" entry
    at tryPackage (node:internal/modules/cjs/loader:444:19)
    at Module._findPath (node:internal/modules/cjs/loader:715:18)
    at Module._resolveFilename (node:internal/modules/cjs/loader:1130:27)
    at Module._load (node:internal/modules/cjs/loader:985:27)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (C:\Users\USER\artv1\index.js:1:19)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32) {
  code: 'MODULE_NOT_FOUND',
  path: 'C:\\Users\\USER\\artv1\\node_modules\\artillery\\package.json',
  requestPath: 'artillery'
}
```

I checked node_modules\artillery\package.json and it does have a main entry of  ` "main": "./lib",`

I tried overwritinhg that with my full local path but it made no difference. ie.  ` "main": "C:/Users/USER/artv1/node_modules/artillery/lib",`

Here's my index.js.  (The detail above `C:\Users\USER\artv1\index.js:1:19` refers to the call below `require('artillery')`)

```
const artillery = require('artillery');

exports.handler = async (event) => {
  try {


	await artillery.run({
		config: require('./loadtest.yml'), // path to your Artillery script
	});
	return { message: "Artillery test executed" };



  } catch (error) {
    console.error(error);

    return {
      statusCode: 500,
      body: JSON.stringify({ message: 'Internal Server Error' }),
    };
  }
};
```

here's my package.json.  I left in the "hello world" stuff because I previously did a simple package to verify everything worked, it only failed once artillery was introduced, ie. main: index.js was correct)

```
{
  "name": "hello-world-nodejs",
  "version": "1.0.0",
  "description": "A simple 'Hello, World!' Node.js application",
  "main": "index.js",
  "scripts": {
    "start": "node index.js"
  },
  "author": "Your Name",
  "license": "MIT",
  "dependencies": {
    "artillery": "^2.0.2"
  }
}
```

I also did a simple test using an axios module to fetch http content and it worked, so there weren't any general issues loading modules, only with artillery.  All those pre-artillery tests also ran fine in my AWS environment using this packaging.

This is on a Windows machine, this was my first Node project on this device.

I've tried removing the global install and uninstalling and reinstalling and updating.  I've tried different versions of Node down to 14 and up to 21.3.0.

thanks, if I'm going the wrong route in order to run on AWS and should pivot away from packaging in Node I can try something else.

**Comment by bernardobridge** [comment]
Hey @timelinedr 👋 ,

Artillery doesn't expose a programmatic API at this time, so you cannot use it directly by requiring it in a NodeJS file. Typically if this is needed, users will call the artillery CLI directly using something like `child_process`, `execa`, `shelljs` or similar.

That being said, if the reason you need this is to use it with Lambda, then we already have built-in [Lambda support](https://www.artillery.io/docs/load-testing-at-scale/aws-lambda) through the [CLI](https://www.artillery.io/docs/reference/cli/run-lambda). The lambda handler we use actually also calls Artillery using [child_process](https://github.com/artilleryio/artillery/blob/main/packages/artillery/lib/platform/aws-lambda/lambda-handler/index.js#L175) as recommended above.

Anyway, all you need to do is use the `run-lambda` command, provided you have the AWS permissions listed in the documentation above correctly set up:

```sh
artillery run-lambda [options] <script>
```

Bear in mind we do currently [have an issue we're investigating](https://github.com/artilleryio/artillery/discussions/2339) on the latest versions (2.0.1, 2.0.2), but we'll fix that soon.

**Comment by mikehalleen** [comment]
Ahhhhhhh, ok thank you.
